code/modules/projectiles/guns/smartgun.dm

```diff
@@ -346,45 +360,37 @@
 		remove_bullet_trait("iff")
 		drain -= 10
 		MD.iff_signal = null
-	if(!powerpack)
-		link_powerpack(usr)
 
 /obj/item/weapon/gun/smartgun/Fire(atom/target, mob/living/user, params, reflex = 0, dual_wield)
-	if(!requires_powerpack)
+	if(!requires_battery)
 		..()
 		return
 
-	if(!powerpack || (powerpack && user.back != powerpack))
-		if(!link_powerpack(user))
-			to_chat(user, SPAN_WARNING("You need a powerpack to be able to fire \the [src]..."))
-			unlink_powerpack()
-			return
-	if(powerpack)
+	if(battery)
 		if(!requires_power)
 			..()
 			return
-		var/obj/item/smartgun_powerpack/pp = user.back
-		if(istype(pp))
-			var/obj/item/cell/c = pp.pcell
-			var/d = drain
-			if(flags_gun_features & GUN_BURST_ON)
-				d = drain*burst_amount*1.5
-			if(pp.drain_powerpack(d, c))
-				..()
-
-
-/obj/item/weapon/gun/smartgun/proc/link_powerpack(mob/user)
-	if(!requires_powerpack)
-		return TRUE
+		if(drain_battery())
+			..()
 
-	if(!QDELETED(user) && !QDELETED(user.back))
-		if(istype(user.back, /obj/item/smartgun_powerpack))
-			powerpack = user.back
-			return TRUE
-	return FALSE
+/obj/item/weapon/gun/smartgun/proc/drain_battery(override_drain)
 
-/obj/item/weapon/gun/smartgun/proc/unlink_powerpack()
-	powerpack = null
+	var/actual_drain = (rand(drain / 2, drain) / 25)
+
+	if(override_drain)
+		actual_drain = (rand(override_drain / 2, override_drain) / 25)
+
+	if(battery && battery.power_cell.charge > 0)
+		if(battery.power_cell.charge > actual_drain)
+			battery.power_cell.charge -= actual_drain
+		else
+			battery.power_cell.charge = 0
+			to_chat(usr, SPAN_WARNING("[src] emits a low power warning and immediately shuts down!"))
```

Comment:
isn't that missing a return FALSE? also it would make the check below for charge==0 redundant